src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs

```diff
@@ -0,0 +1,932 @@
+            hash *= 0x9FB21C651E98DF25;
+            return XorShift(hash, 28);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ulong HashLength9To16(ReadOnlySpan<byte> source, ulong seed)
+        {
+            Debug.Assert(source.Length >= 9 && source.Length <= 16);
+
+            ulong bitflipLow = 0x6782737BEA4239B9 + seed; // DefaultSecretUInt64[3] ^ DefaultSecretUInt64[4]
+            ulong bitflipHigh = 0xAF56BC3B0996523A - seed; // DefaultSecretUInt64[5] ^ DefaultSecretUInt64[6]
+
+            ulong inputLow = BinaryPrimitives.ReadUInt64LittleEndian(source) ^ bitflipLow;
+            ulong inputHigh = BinaryPrimitives.ReadUInt64LittleEndian(source.Slice(source.Length - 8)) ^ bitflipHigh;
+
+            return Avalanche(
+                (ulong)source.Length +
+                BinaryPrimitives.ReverseEndianness(inputLow) +
+                inputHigh +
+                Multiply64To128ThenFold(inputLow, inputHigh));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ulong HashLength17To128(ReadOnlySpan<byte> source, ulong seed)
+        {
+            Debug.Assert(source.Length >= 17 && source.Length <= 128);
+
+            ulong acc = (ulong)source.Length * XxHash64.Prime64_1;
+            if ((uint)source.Length > 32)
+            {
+                if ((uint)source.Length > 64)
+                {
+                    if ((uint)source.Length > 96)
+                    {
+                        acc += Mix16Bytes(source.Slice(48), 0x3F349CE33F76FAA8, 0x1D4F0BC7C7BBDCF9, seed); // DefaultSecretUInt64[12], DefaultSecretUInt64[13]
+                        acc += Mix16Bytes(source.Slice(source.Length - 64), 0x3159B4CD4BE0518A, 0x647378D9C97E9FC8, seed); // DefaultSecretUInt64[14], DefaultSecretUInt64[15]
+                    }
+
+                    acc += Mix16Bytes(source.Slice(32), 0xCB00C391BB52283C, 0xA32E531B8B65D088, seed); // DefaultSecretUInt64[8], DefaultSecretUInt64[9]
+                    acc += Mix16Bytes(source.Slice(source.Length - 48), 0x4EF90DA297486471, 0xD8ACDEA946EF1938, seed); // DefaultSecretUInt64[10], DefaultSecretUInt64[11]
+                }
+
+                acc += Mix16Bytes(source.Slice(16), 0x78E5C0CC4EE679CB, 0x2172FFCC7DD05A82, seed); // DefaultSecretUInt64[4], DefaultSecretUInt64[5]
+                acc += Mix16Bytes(source.Slice(source.Length - 32), 0x8E2443F7744608B8, 0x4C263A81E69035E0, seed); // DefaultSecretUInt64[6], DefaultSecretUInt64[7]
+            }
+
+            acc += Mix16Bytes(source, 0xBE4BA423396CFEB8, 0x1CAD21F72C81017C, seed); // DefaultSecretUInt64[0], DefaultSecretUInt64[1]
+            acc += Mix16Bytes(source.Slice(source.Length - 16), 0xDB979083E96DD4DE, 0x1F67B3B7A4A44072, seed); // DefaultSecretUInt64[2], DefaultSecretUInt64[3]
+
+            return Avalanche(acc);
+        }
+
+        private static ulong HashLength129To240(ReadOnlySpan<byte> source, ulong seed)
```

Comment:
@dotnet/jit-contrib Just FYI, I checked this locally and because of AggressiveInlining everywhere this ends up with 817 locals, pretty close the threshold jit can track (last year we bumped it from 512 to 1024), after crossing which codegen blows up 🙂

Reply:
I based these AggressiveInlinings on the inlining comments / indications in the original C code.  I'm fine to remove some / change use of locals / whatever else.

Reply:
Some motivation to work on #72740 🙂

Reply:
> I based these AggressiveInlinings on the inlining comments / indications in the original C code. I'm fine to remove some / change use of locals / whatever else.

Nah, I was just looking for real-world examples where hit that limit